grouper/fe/handlers.py

```diff
@@ -858,6 +886,172 @@ def get(self, group_id=None, name=None):
+            argument = form.argument_text.data
+
+        permission = Permission.get(self.session, form.permission_name.data)
+        assert permission, "our preseeded permission should exist or we have problems"
+
+        # save off request
+        try:
+            permissions.create_request(self.session, self.current_user, group,
+                    permission, argument, form.reason.data)
+        except permissions.RequestAlreadyGranted:
+            alerts = [Alert("danger", "this group already has this  permission + argument")]
+        except permissions.RequestAlreadyExists:
+            alerts = [Alert("danger", "request for permission + argument already exists")]
+        except permissions.NoOwnersAvailable:
+            alerts = [Alert("danger", "no owners available for requested permission + argument")]
+        else:
+            alerts = None
+
+        if alerts:
+            form = self._prep_form(form, args_by_perm)
+            return self.render(
+                    "group-permission-request.html", form=form, group=group,
+                    args_by_perm_json=json.dumps(args_by_perm),
+                    alerts=alerts,
+                    )
+        else:
+            return self.redirect("/groups/{}".format(group.name))
+
+
+class PermissionsRequests(GrouperHandler):
+    """Allow a user to review a list of permission requests that they have."""
+    def get(self):
+        form = PermissionRequestsForm(self.request.arguments)
+        form.status.choices = [("", "")] + [(k, k) for k in REQUEST_STATUS_CHOICES]
+
+        if not form.validate():
+            alerts = self.get_form_alerts(form.errors)
+            request_tuple = None
+            total = 0
+        else:
+            alerts = []
+            request_tuple, total = permissions.get_requests_by_owner(self.session,
+                    self.current_user, status=form.status.data,
+                    limit=form.limit.data, offset=form.offset.data)
+
+        return self.render("permission-requests.html", form=form, request_tuple=request_tuple,
+                alerts=alerts, total=total, statuses=REQUEST_STATUS_CHOICES)
+
+
+class PermissionsRequestUpdate(GrouperHandler):
+    """Allow a user to action a permisison request they have."""
+    def _get_choices(self, current_status):
+        return [["", ""]] + [
+            [status] * 2
+            for status in REQUEST_STATUS_CHOICES[current_status]
+        ]
+
+    def get(self, request_id):
+        # check for request existence
+        request = permissions.get_request_by_id(self.session, request_id)
+        if not request:
+            return self.notfound()
+
+        # check that this user should be actioning this request
+        user_requests, total = permissions.get_requests_by_owner(self.session,
+                self.current_user, status="pending", limit=None, offset=0)
+        user_request_ids = [ur.id for ur in user_requests.requests]
+        if request.id not in user_request_ids:
```

Comment:
Rather than getting all pending requests for the user and then doing a linear scan, why not just query the database with the above constraints plus the ID?